lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -0,0 +1,257 @@
+//===--- AccessEnforcementDom.cpp - dominated access removal opt ---===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// This function pass removes dynamic access enforcement based on dominance.
+///
+/// General case:
+/// begin_access A (may or may not have no_nested_conflict)
+/// load/store
+/// end_access
+/// ...
+/// begin_access A [no_nested_conflict] // dominated by the first access
+/// load/store
+/// end_access A
+/// The second access scope does not need to be emitted.
+///
+/// Note: This optimization must be aware of all possible access to a Class or
+/// Global address. This includes unpaired access instructions and keypath
+/// entry points. Ignoring any access pattern would weaken enforcement.
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "access-enforcement-dom"
+
+#include "swift/SIL/DebugUtils.h"
+#include "swift/SIL/MemAccessUtils.h"
+#include "swift/SIL/SILFunction.h"
+#include "swift/SILOptimizer/Analysis/DominanceAnalysis.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+
+using namespace swift;
+
+namespace {
+class DominatedAccessRemoval {
+public:
+  using AccessedStoragePair = std::pair<BeginAccessInst *, AccessedStorage>;
+  using AccessedStorageInfo = llvm::SmallVector<AccessedStoragePair, 32>;
+  using DominatorToDominatedPair =
+      std::pair<BeginAccessInst *, BeginAccessInst *>;
+  using DomPairSet = llvm::SmallVector<DominatorToDominatedPair, 32>;
+  using KeyPathEntryPointsSet = llvm::SmallSet<SILInstruction *, 8>;
+
+public:
+  DominatedAccessRemoval(SILFunction &func, DominanceInfo *domInfo)
+      : func(func), domInfo(domInfo) {}
+
+  void perform();
+
+protected:
+  bool visitInstruction(SILInstruction *instr,
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  void visitBeginAccess(BeginAccessInst *beginAccess, AccessedStorage storage,
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  bool domByKeyPath(BeginAccessInst *dominatedInstr);
+  bool analyzeDomSubTree(SILBasicBlock *block,
+                         AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+
+  bool analyze();
+  void optimize();
+
+private:
+  SILFunction &func;
+  DominanceInfo *domInfo;
+  // domPairs is a vector of (Dominating begin_access, Dominated begin_access)
+  // both accesses are dynamic, and the Dominated one has no nested conflict
+  // we can turn the dominated to static during the optimize() phase
+  DomPairSet domPairs;
```

Comment:
This doesn't need to be a pair. You don't care about the first begin_access instruction. It's sufficient to just keep a list of begin_accesses to set the static flag. Also, "Set" is a bad name for it as it is a vector